From 15b97dd730fef1180d70b9706da621a9ae103634 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Tue, 18 Sep 2018 16:27:00 +0800
Subject: [PATCH 6/7] mmc: sdhci-of-esdhc: improve tuning block
 enabling/disabling

According to eSDHC spec, clock should be disabled and ESDHCCTL[FAF]
should be set before tuning block enabling/disabling. This patch
is to move these operations into esdhc_tuning_block_enable() to
reuse them.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
[Xulin: Original patch taken from NXP lx2160a-early-access-bsp0.4-p1.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/mmc/host/sdhci-of-esdhc.c |   13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/mmc/host/sdhci-of-esdhc.c b/drivers/mmc/host/sdhci-of-esdhc.c
index 4694b8e..3ec945f 100644
--- a/drivers/mmc/host/sdhci-of-esdhc.c
+++ b/drivers/mmc/host/sdhci-of-esdhc.c
@@ -737,12 +737,20 @@ static void esdhc_tuning_block_enable(struct sdhci_host *host, bool enable)
 {
 	u32 val;
 
+	esdhc_clock_enable(host, false);
+
+	val = sdhci_readl(host, ESDHC_DMA_SYSCTL);
+	val |= ESDHC_FLUSH_ASYNC_FIFO;
+	sdhci_writel(host, val, ESDHC_DMA_SYSCTL);
+
 	val = sdhci_readl(host, ESDHC_TBCTL);
 	if (enable)
 		val |= ESDHC_TB_EN;
 	else
 		val &= ~ESDHC_TB_EN;
 	sdhci_writel(host, val, ESDHC_TBCTL);
+
+	esdhc_clock_enable(host, true);
 }
 
 static int esdhc_execute_tuning(struct mmc_host *mmc, u32 opcode)
@@ -758,12 +766,7 @@ static int esdhc_execute_tuning(struct mmc_host *mmc, u32 opcode)
 	    host->flags & SDHCI_HS400_TUNING)
 		esdhc_of_set_clock(host, host->clock);
 
-	esdhc_clock_enable(host, false);
-	val = sdhci_readl(host, ESDHC_DMA_SYSCTL);
-	val |= ESDHC_FLUSH_ASYNC_FIFO;
-	sdhci_writel(host, val, ESDHC_DMA_SYSCTL);
 	esdhc_tuning_block_enable(host, true);
-	esdhc_clock_enable(host, true);
 
 	hs400_tuning = host->flags & SDHCI_HS400_TUNING;
 	ret = sdhci_execute_tuning(mmc, opcode);
-- 
1.7.9.5

